Keys

anne-france opened the pull request.

anne-france committed:
package constraint

Commit:
first test for keys

Commit:
correction of test and addition of another tests

aimablehirwa commented:
@blambeau , concernant la classes keys, elle contient une liste de clé candidates ou alors elle doit contenir une relation contenant les valeurs de key pour chaque tuple???
La deuxième solution permet de directement hasher sur les key. L'idée est de faire une projection sur les attributs key pour récupérer la relation contenant que les valeurs de key.

blambeau commented:
Liste de clefs candidates

Envoyé de mon iPhone

> Le 23 mars 2016 à 18:09, aimablehirwa notifications@github.com a écrit :
> 
> @blambeau , concernant la classes keys, elle contient une liste de clé candidates ou alors elle doit contenir une relation contenant les valeurs de key pour chaque tuple???
> La deuxième solution permet de directement hasher sur les key. L'idée est de faire une projection sur les attributs key pour récupérer la relation contenant que les valeurs de key.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub

anne-france committed:
relation pas de clef renvoi tous les attribut clef specifié renvoie clef

anne-france committed:
pas de distinct pour project sur key

anne-france committed:
petite modification ChekKey

anne-france committed:
test Key Equals  hashCode

anne-france committed:
renaming key

Commit:
refactoring and code review of key; many test for key excep for select with key; some test and optimization must be done

aimablehirwa commented:
@blambeau , @anne-france , code review et refactoring

Commit:
test key refactoring

anne-france committed:
Keys

anne-france committed:
correction setkey et checkkey

anne-france committed:
Correction Rename et Test rename

anne-france committed:
correction des operateurs retirer champ key redondant

anne-france committed:
optimisation  de la compilation de minus grace à la key

anne-france committed:
modification select selection pour key plus test

anne-france committed:
correction optimisation compil minus

blambeau commented on src/main/java/jalf/constraint/Constraint.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface Constraint {
+    public boolean Check(Relation r);
+    public AttrList getAttrsKey();
```
`public AttrList getAttrsKey();` makes no sense on Constraint.

blambeau commented on src/main/java/jalf/constraint/ForeignKey.java:
```diff
@@ -0,0 +1,20 @@
+        return null;
+    }
+
+}
```
Ok, mais on va pas tellement utiliser ceci...

blambeau commented on src/main/java/jalf/constraint/Constraint.java:
```diff
@@ -0,0 +1,10 @@
+import jalf.Relation;
+
+public interface Constraint {
+    public boolean Check(Relation r);
```
`check`, pas `Check`

blambeau commented on src/main/java/jalf/constraint/Key.java:
```diff
@@ -0,0 +1,95 @@
+        else{
+            return false;
+        }
+    }
```
A rendre `private` ou carrément implémenter dans `check` plus bas.

blambeau commented on src/main/java/jalf/constraint/Key.java:
```diff
@@ -0,0 +1,95 @@
+                .map(t -> t.project(this.attrsKey, tt))
+                .distinct();
+        return supplier.get().count();
+    }
```
`return r.project(...).cardinality()` ?

blambeau commented on src/main/java/jalf/constraint/Key.java:
```diff
@@ -0,0 +1,95 @@
+    @Override
+    public boolean Check(Relation r) {
+        // TODO Auto-generated method stub
+        return false;
```
j'aurais cherché qqch comme:

```
return this.toAttrList().isSubset(r.getHeading().toAttrList()) && r.cardinality() == r.project(this.toAttrList()).cardinality();
```

blambeau commented on src/main/java/jalf/Relation.java:
```diff
@@ -38,6 +39,16 @@
+     */
+    Key getKey();
+
+    void setKey(Key key);
```
Enlever celle-ci.

blambeau commented on src/main/java/jalf/constraint/Key.java:
```diff
@@ -0,0 +1,95 @@
+    @Override
+    public AttrList getAttrsKey() {
+        return attrsKey;
+    }
```
A renommer `toAttrList()`

blambeau commented on src/main/java/jalf/optimizer/Optimized.java:
```diff
@@ -124,4 +125,16 @@ public boolean sameExpressionAs(Relation other) {
+    @Override
+    public Key getKey() {
+        // TODO Auto-generated method stub
+        return null;
```
`return operator.getKey();`

blambeau commented on src/main/java/jalf/compiler/AbstractRelation.java:
```diff
@@ -30,6 +31,8 @@
 public abstract class AbstractRelation implements Relation {
 
     private TupleType tupleType;
+    private Key key;
```
remove

blambeau commented on src/main/java/jalf/compiler/AbstractRelation.java:
```diff
@@ -93,12 +94,36 @@ public long cardinality() {
+    @Override
+    public Key getKey() {
+        return this.key;
+    }
```
Remove implementation and make the method abstract.

blambeau committed:
Add AttrList.EMPTY

blambeau committed:
Refactor key architecture a little.

- Keep getKey abstract in AbstractRelation
- Implement lazy compute of keys on operator
- Implement base empty key on DUM and DEE
- Enforce key validation on memory relations
- Add known keys to suppliers & parts fixtures
- Fix the tests according to changes above.

anne-france committed:
key se retrouve dans une collection Keys

anne-france committed:
modification structure des tests constraint et  ajout test Keys

Commit:
code review

Commit:
indentation review

Commit:
dsl correction

Commit:
dsl correction

Commit:
code review setmemoryrelation

Commit:
addition of some test

blambeau commented on src/main/java/jalf/constraint/Key.java:
```diff
@@ -0,0 +1,74 @@
+
+    public AttrList intersect(AttrList other){
+	return this.attrsKey.intersect(other);
+    }
```
rename it to `project` instead of `intersect`. Why does it return an AttrList and no a Key?

blambeau commented on src/main/java/jalf/compiler/Cog.java:
```diff
@@ -47,147 +48,157 @@
+	// Key key = keys.toList().get(0);
+	Supplier<Stream<Tuple>> supplier;
+	// avoid duplicate
+	if (keys.stream().anyMatch((Key k) -> k.intersect(on).equals(k))) {
```
Wrong: `k.intersect(on).equals(k)` is always false. k.intersect returns an AttrList while k is a Key => TypeError

The correct semantics is `if k's attributes is a subset of on` then we can optimize by removing the distinct

blambeau replied:
The best, for me, would be to add a `isKeyPreserving(Keys keys)` on Project.

blambeau commented on src/main/java/jalf/constraint/Keys.java:
```diff
@@ -0,0 +1,96 @@
+		    ;
+		});
+	return new Keys(keyunion);
+    }
```
Unclear semantics: it is the cardinal product with the union of key pairs. I doubt it is useful to do that...

We need a simple Set union instead, using addAll twice on a new empty set (at least for relational intersection, possibly join needs the logic above)

blambeau commented on src/main/java/jalf/relation/algebra/Dee.java:
```diff
@@ -31,6 +34,11 @@ public static synchronized Relation instance() {
 
     @Override
+    public Keys getKeys() {
+        return new Keys((Set<Key>) Key.EMPTY);
```
Wrong type cast, a single Key is not a Set<Key>

We need a singleton set with the empty key.

blambeau commented on src/main/java/jalf/relation/algebra/Dum.java:
```diff
@@ -29,6 +32,12 @@ public static synchronized Relation instance() {
+    public Keys getKeys() {
+        return new Keys((Set<Key>) Key.EMPTY);
+        //return Key.EMPTY;
+    }
```
Return a singleton set with the empty key.

blambeau commented on src/main/java/jalf/relation/algebra/Intersect.java:
```diff
@@ -67,4 +70,10 @@ protected RelationType typeCheck() {
+
+    @Override
+    protected Keys lazyComputeKey() {
+        return left.getKeys().union(right.getKeys());
```
ok, provided we use `union({k1,k2},{k3,k4}) = {k1,k2,k3,k4}`, which is not the case right now

blambeau commented on src/main/java/jalf/relation/algebra/Join.java:
```diff
@@ -73,4 +76,10 @@ protected RelationType typeCheck() {
+
+    @Override
+    protected Keys lazyComputeKey() {
+        return left.getKeys().union(right.getKeys());
```
Ok provided we keep union as implemented now.

blambeau commented on src/main/java/jalf/relation/algebra/Project.java:
```diff
@@ -59,4 +67,28 @@ public AttrList getAttributes() {
+            //on a filtrer les mauvaises clefs
+            return new Keys(keyfilter);
+        }
+    }
```
Rephrase and document in terms of new `isKeyPreserving` / `isKeyPreserved` abstractions (methods on Project):

```
isKeyPreserving(Keys keys) -> Boolean
isKeyPreserved(Key key) -> Boolean
```

blambeau commented on src/main/java/jalf/relation/algebra/Union.java:
```diff
@@ -67,4 +69,10 @@ protected RelationType typeCheck() {
+
+    @Override
+    protected Keys lazyComputeKey() {
+        return left.getKeys().union(right.getKeys());
```
No key is preserved: take the largest possible key (the one with all attributes)

blambeau commented on src/main/java/jalf/relation/algebra/Select.java:
```diff
@@ -17,44 +20,49 @@
+    @Override
+    protected Keys lazyComputeKey() {
+	return new Keys(Key.candidate(AttrList.attrs(this.selection.gefns()
+		.keySet())));
```
Add a `getLargestKey(s)` in `AbstractRelation` that delegates to `getLargestKey(s)` on `RelationType` that delegates to `toKey(s)` on `Heading`

blambeau commented on src/main/java/jalf/relation/csv/CsvRelation.java:
```diff
@@ -140,4 +141,10 @@ private CSVReader getReader() throws IOException {
+    @Override
+    public Keys getKeys() {
+        // TODO Auto-generated method stub
+        return null;
```
Use getLargestKey

blambeau commented on src/main/java/jalf/relation/materialized/EmptyRelation.java:
```diff
@@ -43,4 +44,10 @@ public boolean equals(Relation other) {
+    @Override
+    public Keys getKeys() {
+        // TODO Auto-generated method stub
+        return null;
```
return an singleton set with an empty key

Commit:
code review and fixed, two new tests do not succed!

Commit:
check defined keys don't contain a super key of another key, fixation of method toString in keys class, and another test

anne-france committed:
changer branche

aimablehirwa commented:
@blambeau , **attention**, nous avons fini l'implémentation, vous pouvais faire le code review
Nous avons ajouter une méthode pour vérifier si dans 'keys' il n'y pas de clé qui est super clé d'une autre clé.

anne-france committed:
espace et revieuw
